File: server/app/services/security_service.py
import asyncio
import hashlib
import logging
import secrets
import time
from collections import defaultdict
from collections.abc import Callable
from datetime import datetime, timedelta

from fastapi import Request
from fastapi.responses import JSONResponse

logger = logging.getLogger(__name__)

class RateLimiter:
    def __init__(
        self,
        requests_per_minute: int = 60,
        requests_per_hour: int = 1000,
        burst_size: int = 10
    ):
        self.requests_per_minute = requests_per_minute
        self.requests_per_hour = requests_per_hour
        self.burst_size = burst_size
        self._requests: dict[str, list] = defaultdict(list)
        self._lock = asyncio.Lock()

    def _get_client_key(self, request: Request) -> str:
        forwarded = request.headers.get("X-Forwarded-For")
        if forwarded:
            return forwarded.split(",")[0].strip()

        real_ip = request.headers.get("X-Real-IP")
        if real_ip:
            return real_ip

        if request.client:
            return request.client.host

        return "unknown"

    async def is_allowed(self, request: Request) -> tuple:
        key = self._get_client_key(request)
        now = time.time()

        async with self._lock:
            requests = self._requests[key]

            requests[:] = [t for t in requests if now - t < 3600]

            minute_ago = now - 60
            minute_requests = sum(1 for t in requests if t > minute_ago)

            hour_requests = len(requests)

            if minute_requests >= self.requests_per_minute:
                retry_after = int(60 - (now - min(t for t in requests if t > minute_ago)))
                return False, f"超过每分钟请求限制({self.requests_per_minute})", retry_after

            if hour_requests >= self.requests_per_hour:
                retry_after = int(3600 - (now - requests[0]))
                return False, f"超过每小时请求限制({self.requests_per_hour})", retry_after

            requests.append(now)
            return True, None, 0

    async def __call__(self, request: Request, call_next: Callable):
        allowed, message, retry_after = await self.is_allowed(request)

        if not allowed:
            return JSONResponse(
                status_code=429,
                content={
                    "detail": "请求过于频繁",
                    "message": message,
                    "retry_after": retry_after
                },
                headers={"Retry-After": str(retry_after)}
            )

        return await call_next(request)


class SecurityHeaders:
    def __init__(self):
        self.headers = {
            "X-Content-Type-Options": "nosniff",
            "X-Frame-Options": "DENY",
            "X-XSS-Protection": "1; mode=block",
            "Strict-Transport-Security": "max-age=31536000; includeSubDomains",
            "Content-Security-Policy": "default-src 'self'; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'; img-src 'self' data: blob:; font-src 'self' data:;",
            "Referrer-Policy": "strict-origin-when-cross-origin",
            "Permissions-Policy": "geolocation=(), microphone=(), camera=()"
        }

    async def __call__(self, request: Request, call_next: Callable):
        response = await call_next(request)

        for header, value in self.headers.items():
            response.headers[header] = value

        return response


class InputValidator:
    SQL_KEYWORDS = [
        "SELECT", "INSERT", "UPDATE", "DELETE", "DROP", "UNION",
        "OR", "AND", "WHERE", "FROM", "INTO", "VALUES", "SET"
    ]

    XSS_PATTERNS = [
        "<script", "</script>", "javascript:", "onerror=", "onload=",
        "eval(", "document.", "window.", "alert("
    ]

    @classmethod
    def sanitize_string(cls, value: str) -> str:
        if not isinstance(value, str):
            return value

        sanitized = value.strip()

        for pattern in cls.XSS_PATTERNS:
            sanitized = sanitized.replace(pattern.lower(), "")
            sanitized = sanitized.replace(pattern.upper(), "")

        return sanitized

    @classmethod
    def check_sql_injection(cls, value: str) -> bool:
        if not isinstance(value, str):
            return False

        upper_value = value.upper()

        for keyword in cls.SQL_KEYWORDS:
            if keyword in upper_value and ("'" in value or '"' in value or ";" in value):
                return True

        return False

    @classmethod
    def validate_file_type(cls, filename: str, allowed_types: list) -> bool:
        if not filename:
            return False

        ext = filename.lower().split(".")[-1] if "." in filename else ""
        return ext in allowed_types


class CSRFProtection:
    def __init__(self, secret_key: str | None = None):
        self.secret_key = secret_key or secrets.token_urlsafe(32)
        self._tokens: dict[str, datetime] = {}

    def generate_token(self, session_id: str) -> str:
        timestamp = str(time.time())
        data = f"{session_id}:{timestamp}:{self.secret_key}"
        token = hashlib.sha256(data.encode()).hexdigest()

        self._tokens[token] = datetime.now() + timedelta(hours=24)

        return token

    def validate_token(self, token: str, session_id: str) -> bool:
        if not token or token not in self._tokens:
            return False

        if datetime.now() > self._tokens[token]:
            del self._tokens[token]
            return False

        return True

    def cleanup_expired(self):
        now = datetime.now()
        expired = [t for t, exp in self._tokens.items() if now > exp]
        for token in expired:
            del self._tokens[token]


class SecurityMiddleware:
    def __init__(
        self,
        rate_limit_per_minute: int = 60,
        rate_limit_per_hour: int = 1000
    ):
        self.rate_limiter = RateLimiter(
            requests_per_minute=rate_limit_per_minute,
            requests_per_hour=rate_limit_per_hour
        )
        self.security_headers = SecurityHeaders()
        self.csrf = CSRFProtection()

    async def __call__(self, request: Request, call_next: Callable):
        response = await self.rate_limiter(request, call_next)

        if isinstance(response, JSONResponse) and response.status_code == 429:
            return response

        for header, value in self.security_headers.headers.items():
            response.headers[header] = value

        return response

    def get_csrf_token(self, session_id: str) -> str:
        return self.csrf.generate_token(session_id)

    def validate_csrf(self, token: str, session_id: str) -> bool:
        return self.csrf.validate_token(token, session_id)


security_middleware = SecurityMiddleware()

File: apps/api/src/services/security-service.ts
/**
 * 安全服务（风控/异常检测/IP 信誉/输入校验）。
 * 迁移自旧架构 security_service.py。
 *
 * 提供：
 * - RateLimiter: 基于 Redis 滑动窗口限流（每分钟/每小时）
 * - InputValidator: XSS 清洗、SQL 注入检测、文件类型校验
 * - CSRFProtection: CSRF token 生成与校验
 * - SECURITY_HEADERS: 安全响应头
 */

import { createHash, randomBytes } from 'node:crypto';
import type { FastifyRequest } from 'fastify';
import type IORedis from 'ioredis';

// =============================================================================
// RateLimiter - Redis 滑动窗口限流
// =============================================================================

export interface RateLimitResult {
  allowed: boolean;
  message?: string;
  retryAfter?: number;
}

export class RateLimiter {
  constructor(
    private readonly redis: IORedis,
    private readonly requestsPerMinute = 60,
    private readonly requestsPerHour = 1000,
  ) {}

  /** 从请求中提取客户端 IP(2026-07-21 安全审计第十轮加固)。
   *
   * 不再直接读 X-Forwarded-For 头 — 因为 Fastify 已通过 trustProxy 配置
   * 在解析阶段验证代理 IP,request.ip 已经是真实客户端 IP(或最后可信代理 IP)。
   * 手工读 X-Forwarded-For 会绕过 trustProxy 检查,被攻击者伪造任意 IP 绕过
   * 限流 / IP 拉黑 / 失败计数(例如:Authorization: bypass, X-Forwarded-For: 1.1.1.1)。
   */
  static getClientIp(request: FastifyRequest): string {
    return request.ip
  }

  /**
   * 检查是否允许请求。
   * key 维度由调用方决定（IP / userId / IP+路由）。
   */
  async isAllowed(key: string): Promise<RateLimitResult> {
    const minuteKey = `rl:m:${key}`;
    const hourKey = `rl:h:${key}`;

    const [minuteCount, hourCount] = await Promise.all([
      this.redis.incr(minuteKey),
      this.redis.incr(hourKey),
    ]);
    // 首次写入时设置 TTL
    if (minuteCount === 1) await this.redis.expire(minuteKey, 60);
    if (hourCount === 1) await this.redis.expire(hourKey, 3600);

    if (minuteCount > this.requestsPerMinute) {
      return {
        allowed: false,
        message: `超过每分钟请求限制(${this.requestsPerMinute})`,
        retryAfter: 60,
      };
    }
    if (hourCount > this.requestsPerHour) {
      return {
        allowed: false,
        message: `超过每小时请求限制(${this.requestsPerHour})`,
        retryAfter: 3600,
      };
    }
    return { allowed: true };
  }
}

// =============================================================================
// InputValidator - 输入校验
// =============================================================================

const SQL_KEYWORDS = [
  'SELECT', 'INSERT', 'UPDATE', 'DELETE', 'DROP', 'UNION',
  'OR', 'AND', 'WHERE', 'FROM', 'INTO', 'VALUES', 'SET',
];

const XSS_PATTERNS = [
  '<script', '</script>', 'javascript:', 'onerror=', 'onload=',
  'eval(', 'document.', 'window.', 'alert(',
];

export class InputValidator {
  /** 清洗 XSS（移除常见脚本标签与事件处理器）。 */
  static sanitizeString(value: string): string {
    if (typeof value !== 'string') return value;
    let sanitized = value.trim();
    for (const pattern of XSS_PATTERNS) {
      const re = new RegExp(pattern.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'gi');
      sanitized = sanitized.replace(re, '');
    }
    return sanitized;
  }

  /** 检测 SQL 注入（关键字 + 引号/分号组合）。 */
  static checkSqlInjection(value: string): boolean {
    if (typeof value !== 'string') return false;
    const upper = value.toUpperCase();
    const hasQuote = /['";]/.test(value);
    if (!hasQuote) return false;
    return SQL_KEYWORDS.some((kw) => upper.includes(kw));
  }

  /** 校验文件扩展名是否在允许列表中。 */
  static validateFileType(filename: string, allowedTypes: string[]): boolean {
    if (!filename) return false;
    const ext = filename.toLowerCase().split('.').pop() ?? '';
    return allowedTypes.includes(ext);
  }
}

// =============================================================================
// CSRFProtection - CSRF token 生成与校验（Redis 存储）
// =============================================================================

export class CSRFProtection {
  private readonly prefix = 'csrf:';

  constructor(private readonly redis: IORedis, private readonly ttlSeconds = 86400) {}

  /** 生成 CSRF token 并存入 Redis（绑定 sessionId）。 */
  async generateToken(sessionId: string): Promise<string> {
    const token = randomBytes(32).toString('hex');
    const data = `${sessionId}:${token}`;
    const key = `${this.prefix}${createHash('sha256').update(data).digest('hex')}`;
    await this.redis.setex(key, this.ttlSeconds, sessionId);
    return token;
  }

  /** 校验 CSRF token。 */
  async validateToken(token: string, sessionId: string): Promise<boolean> {
    if (!token) return false;
    const data = `${sessionId}:${token}`;
    const key = `${this.prefix}${createHash('sha256').update(data).digest('hex')}`;
    const stored = await this.redis.get(key);
    return stored === sessionId;
  }

  /** 清理过期 token（Redis TTL 自动清理，此方法为兼容保留）。 */
  async cleanupExpired(): Promise<void> {
    // Redis setex 自动过期，无需手动清理
  }
}

// =============================================================================
// 安全响应头
// =============================================================================

export const SECURITY_HEADERS: Record<string, string> = {
  'X-Content-Type-Options': 'nosniff',
  'X-Frame-Options': 'DENY',
  'X-XSS-Protection': '1; mode=block',
  'Strict-Transport-Security': 'max-age=31536000; includeSubDomains',
  'Content-Security-Policy':
    "default-src 'self'; script-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'; img-src 'self' data: blob:; font-src 'self' data:;",
  'Referrer-Policy': 'strict-origin-when-cross-origin',
  'Permissions-Policy': 'geolocation=(), microphone=(), camera=()',
};

/**
 * 异常检测：检查 IP 是否在黑名单（Redis set）中。
 * 黑名单 key: sec:ip:blacklist
 */
export async function isIpBlacklisted(redis: IORedis, ip: string): Promise<boolean> {
  try {
    return (await redis.sismember('sec:ip:blacklist', ip)) === 1;
  } catch {
    return false;
  }
}

/**
 * IP 信誉记录：连续失败 N 次后自动加入黑名单。
 * key: sec:ip:fail:<ip>（计数），达到阈值则写入黑名单。
 */
export async function recordIpFailure(
  redis: IORedis,
  ip: string,
  threshold = 10,
  windowSeconds = 3600,
): Promise<boolean> {
  const key = `sec:ip:fail:${ip}`;
  const count = await redis.incr(key);
  if (count === 1) await redis.expire(key, windowSeconds);
  if (count >= threshold) {
    await redis.sadd('sec:ip:blacklist', ip);
    return true;
  }
  return false;
}

/** 清除 IP 失败计数（登录成功时调用）。 */
export async function clearIpFailures(redis: IORedis, ip: string): Promise<void> {
  await redis.del(`sec:ip:fail:${ip}`);
}
